File: go/protocol.go
package plutus

import (
	"fmt"
	"net/url"
	"sort"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

// ProtocolProfile selects a wire contract explicitly. Zero value preserves Alpha.
type ProtocolProfile string

const (
	RequestBoundV1 ProtocolProfile = "request-bound-v1"
	ProductV1      ProtocolProfile = "product-v1"
)

func (p ProtocolProfile) validate() error {
	if p != "" && p != RequestBoundV1 && p != ProductV1 {
		return fmt.Errorf("%w: unknown protocol profile", ErrInvalidConfig)
	}
	return nil
}

// CanonicalizeQueryForProfile uses the selected protocol without fallback.
func CanonicalizeQueryForProfile(raw string, profile ProtocolProfile) (string, error) {
	if err := profile.validate(); err != nil {
		return "", err
	}
	if profile != ProductV1 {
		return CanonicalizeQuery(raw)
	}
	if strings.TrimFunc(raw, javaWhitespace) == "" {
		return "", nil
	}
	pairs := make([][2]string, 0)
	for _, part := range strings.Split(raw, "&") {
		key, value, _ := strings.Cut(part, "=")
		k, err := url.QueryUnescape(key)
		if err != nil || !utf8.ValidString(k) {
			return "", fmt.Errorf("%w: invalid product query key", ErrInvalidConfig)
		}
		v, err := url.QueryUnescape(value)
		if err != nil || !utf8.ValidString(v) {
			return "", fmt.Errorf("%w: invalid product query value", ErrInvalidConfig)
		}
		pairs = append(pairs, [2]string{k, v})
	}
	sort.SliceStable(pairs, func(i, j int) bool {
		if pairs[i][0] != pairs[j][0] {
			return utf16Less(pairs[i][0], pairs[j][0])
		}
		return utf16Less(pairs[i][1], pairs[j][1])
	})
	out := make([]string, len(pairs))
	encode := func(value string) string {
		return strings.NewReplacer("+", "%20", "%2A", "*", "~", "%7E").Replace(url.QueryEscape(value))
	}
	for i, pair := range pairs {
		out[i] = encode(pair[0]) + "=" + encode(pair[1])
	}
	return strings.Join(out, "&"), nil
}

func javaWhitespace(r rune) bool {
	return r >= 9 && r <= 13 || r >= 0x1c && r <= 0x20 || r == 0x1680 || r >= 0x2000 && r <= 0x2006 || r >= 0x2008 && r <= 0x200a || r == 0x2028 || r == 0x2029 || r == 0x205f || r == 0x3000
}
func utf16Less(a, b string) bool {
	x, y := utf16.Encode([]rune(a)), utf16.Encode([]rune(b))
	for i := 0; i < len(x) && i < len(y); i++ {
		if x[i] != y[i] {
			return x[i] < y[i]
		}
	}
	return len(x) < len(y)
}
func productRequestID(value string) (string, error) {
	if strings.ContainsAny(value, "\r\n") {
		return "", fmt.Errorf("%w: request ID must not contain newlines", ErrInvalidConfig)
	}
	if id := strings.TrimSpace(value); id != "" {
		return id, nil
	}
	nonce, err := NewNonce()
	if err != nil {
		return "", err
	}
	return "req_" + nonce, nil
}

File: node/src/protocol.ts
import { randomBytes } from 'node:crypto';
import { PlutusConfigError, PlutusCanonicalizationError, PlutusRequestError } from './errors.js';

export enum ProtocolProfile {
  REQUEST_BOUND_V1 = 'request-bound-v1',
  PRODUCT_V1 = 'product-v1',
}

export function resolveProfile(value: ProtocolProfile = ProtocolProfile.REQUEST_BOUND_V1): ProtocolProfile {
  if (value !== ProtocolProfile.REQUEST_BOUND_V1 && value !== ProtocolProfile.PRODUCT_V1) {
    throw new PlutusConfigError('unknown protocolProfile');
  }
  return value;
}

export function productCanonicalQuery(raw: string | null | undefined): string {
  if (raw == null || /^[\u0009-\u000d\u001c-\u0020\u1680\u2000-\u2006\u2008-\u200a\u2028\u2029\u205f\u3000]*$/u.test(raw)) return '';
  try {
    const encode = (value: string): string => encodeURIComponent(value).replace(/[!'()~]/g, ch => '%' + ch.charCodeAt(0).toString(16).toUpperCase());
    const decode = (value: string): string => decodeURIComponent(value.replace(/\+/g, ' '));
    const pairs = raw.split('&').map(part => {
      const index = part.indexOf('=');
      return [decode(index < 0 ? part : part.slice(0, index)), decode(index < 0 ? '' : part.slice(index + 1))] as const;
    });
    // Java String ordering compares decoded UTF-16 code units, as JavaScript does.
    pairs.sort((a, b) => a[0] !== b[0] ? (a[0] < b[0] ? -1 : 1) : a[1] === b[1] ? 0 : a[1] < b[1] ? -1 : 1);
    return pairs.map(([key, value]) => encode(key) + '=' + encode(value)).join('&');
  } catch {
    throw new PlutusCanonicalizationError('invalid product query encoding');
  }
}

export function productRequestId(value?: string | null): string {
  if (value != null && /[\r\n]/.test(value)) throw new PlutusRequestError('X-Request-Id must not contain newlines');
  return value?.trim() || 'req_' + randomBytes(16).toString('hex');
}
